File: Sources/structly/04_linked_list/ReverseList.swift
//
//  ReverseList.swift
//  MyCLI
//
//  Created by new on 3/18/26.
//

func reverseList(_ head: Node<String>?) -> Node<String>? {
    func rev(node: Node<String>?, prevNode: Node<String>?) -> Node<String>? {
        guard let node else {
            return prevNode
        }

        let newHead = rev(node: node.next, prevNode: node)
        node.next = prevNode

        return newHead
    }

    return rev(node: head, prevNode: nil)
}

func reverseList1(_ head: Node<String>?) -> Node<String>? {
    var prevNode: Node<String>? = nil
    var node: Node<String>? = head
    while let currentNode = node {
        let nextNode = currentNode.next
        currentNode.next = prevNode
        prevNode = currentNode
        node = nextNode
    }
    
    return prevNode
}

File: Sources/structlyC/04_linked_list/reverse_list.cpp
//
//  reverse_linked_list.h
//  MyCLI
//
//  Created by new on 3/18/26.
//

#include <string>

class Node {
public:
    std::string val;
    Node* next;

    Node(std::string initialVal) {
        val = initialVal;
        next = nullptr;
    }
};

Node* rev_helper(Node* node, Node* prev) {
    if (node == nullptr) {
        return prev;
    }

    Node* newHead = rev_helper(node->next, node);
    node->next = prev;

    return newHead;
}

Node* reverseList(Node* head) {
    return rev_helper(head, nullptr);
}

Node* reverseList1(Node* head) {
    Node* node = head;
    Node* node_prev = nullptr;
    while (node != nullptr) {
        Node* tmp = node->next;
        node->next = node_prev;
        node_prev = node;
        node = tmp;
    }

    return node_prev;
}
